### PR TITLE
Certified data prop test

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    /// as in these cases the system already certifies the response)
+    @query(
+        [],
+        IDL.Record({
+            value: IDL.Nat32,
+            certificate: IDL.Opt(IDL.Vec(IDL.Nat8))
+        })
+    )
+    async get(): Promise<{ value: number; certificate: [Uint8Array] | [] }> {
+        const cert = dataCertificate();
```
Comment (bdemann) [inline tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/src/index.ts:38]: lets go ahead and rename this certificate

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                    }
+                ),
+                defaultPropTestParams()
+            );
+        });
+    };
+}
+
+async function testCertifiedVariableUpdate(
+    newVal: number,
```
Comment (bdemann) [inline tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts:51]: new value?

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    agent: HttpAgent,
+    canisterName: string
+): Promise<void> {
+    const canisterPrincipal = Principal.fromText(getCanisterId(canisterName));
+
+    await certVarCanister.set(newVal);
+
+    const result = await certVarCanister.get();
+
+    const cert = await verifyCertificate(
```
Comment (bdemann) [inline tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts:62]: certificate?

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    canisterPrincipal: Principal
+): Promise<Certificate> {
+    if (certificateBytes.length === 0) {
+        throw new Error('No certificate found');
+    }
+    expect(certificateBytes.length).toBeGreaterThan(0);
+
+    const certificate = new Uint8Array(certificateBytes[0]).buffer;
+
+    const cert = await Certificate.create({
```
Comment (bdemann) [inline tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts:83]: certificate

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        certificate,
+        rootKey: agent.rootKey,
+        canisterId: canisterPrincipal
+    });
+
+    return cert;
+}
+
+function verifyTimestamp(cert: Certificate): void {
+    const certTime = findLookupValueOrThrow(cert, ['time']);
```
Comment (bdemann) [inline tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts:93]: certificateTime

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    const time = Number(decodedTime) / 1e9;
+
+    const now = Date.now() / 1000;
+    const diff = Math.abs(time - now);
+
+    expect(diff).toBeLessThan(5);
+    // Timestamp difference seems legit (< 5 sec)
+}
+
+// This function verifies the certified data in the certificate.
```
Comment (bdemann) [inline tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts:104]: I don't think we need this comment

File: tests/property/ic_api/certified_data/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    init,
+    postUpgrade,
+    preUpgrade,
+    query,
+    setCertifiedData,
+    StableBTreeMap,
+    update
+} from 'azle';
+
+let afterFirstPostUpgrade = false;
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/src/index.ts:13]: make this a class member

File: tests/property/ic_api/certified_data/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            setCertifiedData(PRE_UPGRADE_DATA);
+            this.stableStorage.insert('certifiedDataSetInPreUpgrade', true);
+        }
+    }
+
+    @postUpgrade([IDL.Bool, IDL.Vec(IDL.Nat8)])
+    postUpgrade(setData: boolean, data: Uint8Array): void {
+        const wasSetInPreUpgrade = this.stableStorage.get(
+            'certifiedDataSetInPreUpgrade'
+        );
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/src/index.ts:45]: add space here

File: tests/property/ic_api/certified_data/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        // We could make it arbitrary but that seems like a lot of work just to test this one case
+        if (afterFirstPostUpgrade) {
+            setCertifiedData(PRE_UPGRADE_DATA);
+            this.stableStorage.insert('certifiedDataSetInPreUpgrade', true);
+        }
+    }
+
+    @postUpgrade([IDL.Bool, IDL.Vec(IDL.Nat8)])
+    postUpgrade(setData: boolean, data: Uint8Array): void {
+        const wasSetInPreUpgrade = this.stableStorage.get(
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/src/index.ts:43]: make this the full name

File: tests/property/ic_api/certified_data/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    @update([], IDL.Opt(IDL.Vec(IDL.Nat8)))
+    getDataCertificateInUpdate(): [Uint8Array] | [] {
+        const certificate = dataCertificate();
+        return certificate ? [certificate] : [];
+    }
+
+    @query([])
+    setDataCertificateInQuery(): void {
+        try {
+            setCertifiedData(new Uint8Array([3]));
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/src/index.ts:90]: this try doesn't do anything... it still just errors out...

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 11]
```diff
@@ -0,0 +1,413 @@
...
+                            CANISTER_NAME
+                        );
+
+                        deployCanister({
+                            setData: true,
+                            force: true,
+                            initData
+                        });
+
+                        // Check that getCertificate returns [0] after upgrade
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/test/tests.ts:158]: get rid of comment

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 12]
```diff
@@ -0,0 +1,413 @@
...
+                new Uint8Array(),
+                actor,
+                agent,
+                CANISTER_NAME
+            );
+
+            // Upgrade the canister
+            deployCanister({ force: true });
+
+            // Check that getCertificate returns an empty array initially
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/test/tests.ts:194]: after first post upgrade is what that comment should say

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+            await testCertifiedData(
+                new Uint8Array(),
+                actor,
+                agent,
+                CANISTER_NAME
+            );
+
+            deployCanister({ force: true });
+
+            // Check that getCertificate returns [0] after upgrade
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/test/tests.ts:204]: remove comment

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 14]
```diff
@@ -0,0 +1,413 @@
...
+            await testCertifiedData(
+                preUpgradeBytes,
+                actor,
+                agent,
+                CANISTER_NAME
+            );
+        });
+
+        it('always returns empty array when trying to get data certificate in update method', async () => {
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/test/tests.ts:213]: always return []

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 15]
```diff
@@ -0,0 +1,413 @@
...
+    canisterPrincipal: Principal
+): Promise<Certificate> {
+    if (certificateBytes.length === 0) {
+        throw new Error('No certificate found');
+    }
+    expect(certificateBytes.length).toBeGreaterThan(0);
+
+    const certificate = new Uint8Array(certificateBytes[0]).buffer;
+
+    const cert = await Certificate.create({
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/test/tests.ts:366]: return here

Comment (bdemann) [reply tests/property/ic_api/certified_data/test/tests.ts:366]: also in the other place?

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 16]
```diff
@@ -0,0 +1,413 @@
...
+        certificate,
+        rootKey: agent.rootKey,
+        canisterId: canisterPrincipal
+    });
+
+    return cert;
+}
+
+function verifyTimestamp(cert: Certificate): void {
+    const certTime = findLookupValueOrThrow(cert, ['time']);
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/test/tests.ts:376]: certificateTime

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+    const cert = await Certificate.create({
+        certificate,
+        rootKey: agent.rootKey,
+        canisterId: canisterPrincipal
+    });
+
+    return cert;
+}
+
+function verifyTimestamp(cert: Certificate): void {
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/test/tests.ts:375]: certificate

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+
+    expect(diff).toBeLessThan(5);
+    // Timestamp difference seems legit (< 5 sec)
+}
+
+// This function verifies the certified data in the certificate.
+// Consequently it also verifies the canister ID and the value.
+function verifyCertifiedData(
+    cert: Certificate,
+    cid: Principal,
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/test/tests.ts:391]: canisterPrincipal

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+    const candidEncodedRawData = new Uint8Array([
+        ...new TextEncoder().encode('DIDL\x01\x6d\x7b\x01\x00'),
+        ...new TextEncoder().encode(lengthByte),
+        ...new Uint8Array(rawData)
+    ]);
+
+    const decodedData = IDL.decode(
+        [IDL.Vec(IDL.Nat8)],
+        candidEncodedRawData
+    )[0];
```
Comment (bdemann) [inline tests/property/ic_api/certified_data/test/tests.ts:411]: new line

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 20]
```diff
@@ -0,0 +1,415 @@
...
+                        });
+                        const agent = await createAuthenticatedAgent(whoami());
+                        const actor = await getCanisterActor<Actor>(
+                            CANISTER_NAME,
+                            {
+                                agent
+                            }
+                        );
+
+                        // Check that getCertificate returns [0] initially
```
Comment (lastmjs) [inline tests/property/ic_api/certified_data/test/tests.ts:116]: Get rid of this comment

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 21]
```diff
@@ -0,0 +1,415 @@
...
+                ),
+                defaultPropTestParams()
+            );
+        });
+
+        it('sets certified data in init', async () => {
+            await fc.assert(
+                fc.asyncProperty(
+                    fc.uint8Array({ minLength: 1, maxLength: 32 }),
+                    async (initData) => {
```
Comment (lastmjs) [inline tests/property/ic_api/certified_data/test/tests.ts:102]: Just call initData `data`

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 22]
```diff
@@ -0,0 +1,415 @@
...
+                ),
+                defaultPropTestParams()
+            );
+        });
+
+        it('sets certified data in post upgrade', async () => {
+            await fc.assert(
+                fc.asyncProperty(
+                    fc.uint8Array({ minLength: 1, maxLength: 32 }),
+                    async (initData) => {
```
Comment (lastmjs) [inline tests/property/ic_api/certified_data/test/tests.ts:133]: Just call initData `data`

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 23]
```diff
@@ -0,0 +1,415 @@
...
+
+            await testCertifiedData(
+                preUpgradeBytes,
+                actor,
+                agent,
+                CANISTER_NAME
+            );
+        });
+
+        it('always returns [] when trying to get data certificate in update method', async () => {
```
Comment (lastmjs) [inline tests/property/ic_api/certified_data/test/tests.ts:212]: clarify comment

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 24]
```diff
@@ -0,0 +1,415 @@
...
+            await fc.assert(
+                fc.asyncProperty(
+                    fc.uint8Array({ minLength: 1, maxLength: 32 }),
+                    async (data) => {
+                        const actor =
+                            await getCanisterActor<Actor>(CANISTER_NAME);
+
+                        await expect(
```
Comment (lastmjs) [inline tests/property/ic_api/certified_data/test/tests.ts:220]: Let's not do this resolves thing

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/pretest.ts
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/pretest.ts:1]: Done

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+        it('verifies certified variable increments', async () => {
+            const canisterName = 'cert-var';
+            const agent = await createAuthenticatedAgent(whoami());
+            const certVarCanister = await getCanisterActor<Actor>(
+                canisterName,
+                {
+                    agent
+                }
+            );
+            await testCertifiedVariableUpdate(
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts:32]: Done

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+            const agent = await createAuthenticatedAgent(whoami());
+            const certVarCanister = await getCanisterActor<Actor>(
+                canisterName,
+                {
+                    agent
+                }
+            );
+            await fc.assert(
+                fc.asyncProperty(
+                    fc.nat({ max: 4294967295 }), // max value for Nat32
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts:60]: Done

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+    const certificate = await verifyCertificate(
+        result.certificate,
+        agent,
+        canisterPrincipal
+    );
+    verifyTimestamp(certificate);
+    verifyCertifiedData(certificate, canisterPrincipal, result.value);
+}
+
+async function verifyCertificate(
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts:98]: Done

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/test/tests.ts:1]: Done

File: tests/property/ic_api/certified_data/src/index.ts [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+]);
+
+export default class {
+    data: Uint8Array = new Uint8Array();
+    stableStorage = StableBTreeMap<string, boolean>(0);
+    afterFirstPostUpgrade = false;
+
+    @init([IDL.Bool, IDL.Vec(IDL.Nat8)])
+    init(setData: boolean, data: Uint8Array): void {
+        if (setData) {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/src/index.ts:25]: Done

File: tests/property/ic_api/certified_data/src/index.ts [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    }
+
+    @postUpgrade([IDL.Bool, IDL.Vec(IDL.Nat8)])
+    postUpgrade(setData: boolean, data: Uint8Array): void {
+        const certifiedDataSetInPreUpgrade = this.stableStorage.get(
+            'certifiedDataSetInPreUpgrade'
+        );
+
+        if (certifiedDataSetInPreUpgrade) {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/src/index.ts:47]: done

File: tests/property/ic_api/certified_data/src/index.ts [hunk 30]
```diff
@@ -0,0 +1,101 @@
...
+            this.data = PRE_UPGRADE_DATA;
+        }
+
+        if (setData) {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/src/index.ts:51]: Did return instead

File: tests/property/ic_api/certified_data/src/index.ts [hunk 31]
```diff
@@ -0,0 +1,101 @@
...
+            this.data = data;
+            setCertifiedData(data);
+        }
+
+        this.afterFirstPostUpgrade = true;
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/src/index.ts:56]: done

File: tests/property/ic_api/certified_data/src/index.ts [hunk 32]
```diff
@@ -0,0 +1,101 @@
...
+
+    @update([IDL.Vec(IDL.Nat8)])
+    setData(data: Uint8Array): void {
+        this.data = data;
+        setCertifiedData(data);
+    }
+
+    @query(
+        [],
+        IDL.Record({
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/src/index.ts:67]: Done

File: tests/property/ic_api/certified_data/test/certificate.ts
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/test/certificate.ts:1]: Done

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 33]
```diff
@@ -0,0 +1,415 @@
...
+} from 'azle/test';
+import { execSync } from 'child_process';
+import fc from 'fast-check';
+
+import { findLookupValueOrThrow } from './certificate';
+import { _SERVICE as Actor } from './dfx_generated/canister/canister.did';
+
+type DeployCanisterOptions = {
+    setData?: boolean;
+    initData?: Uint8Array;
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/test/tests.ts:20]: Done

File: tests/property/ic_api/certified_data/test/tests.ts (hunk 21 above)
Comment (bdemann) [reply tests/property/ic_api/certified_data/test/tests.ts:102]: done

File: tests/property/ic_api/certified_data/test/tests.ts (hunk 20 above)
Comment (bdemann) [reply tests/property/ic_api/certified_data/test/tests.ts:116]: done

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 34]
```diff
@@ -0,0 +1,415 @@
...
+};
+
+const CANISTER_NAME = 'canister';
+
+export function getTests(): Test {
+    return () => {
+        it('sets and gets certified data', async () => {
+            await fc.assert(
+                fc.asyncProperty(
+                    fc.uint8Array({ minLength: 1, maxLength: 32 }),
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/test/tests.ts:31]: done

File: tests/property/ic_api/certified_data/test/tests.ts (hunk 22 above)
Comment (bdemann) [reply tests/property/ic_api/certified_data/test/tests.ts:133]: done

File: tests/property/ic_api/certified_data/test/tests.ts (hunk 23 above)
Comment (bdemann) [reply tests/property/ic_api/certified_data/test/tests.ts:212]: Done

File: tests/property/ic_api/certified_data/test/tests.ts (hunk 24 above)
Comment (bdemann) [reply tests/property/ic_api/certified_data/test/tests.ts:220]: done

File: .eslintrc.cjs
Comment (lastmjs) [inline .eslintrc.cjs:1]: Put this in the linting wishlist, it's too dangerous to be kept alive

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/src/index.ts [hunk 35]
```diff
@@ -1,11 +1,14 @@
...
-    update
-} from 'azle';
+import { dataCertificate, IDL, query, setCertifiedData, update } from 'azle';
+
+type CertifiedValue = {
+    value: number;
+    certificate: [Uint8Array] | [];
+};
+
+const CertifiedValue = IDL.Record({
```
Comment (lastmjs) [inline tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/src/index.ts]: Move to top

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/src/index.ts [hunk 36]
```diff
@@ -1,11 +1,14 @@
...
-    dataCertificate,
-    IDL,
-    query,
-    setCertifiedData,
-    update
-} from 'azle';
+import { dataCertificate, IDL, query, setCertifiedData, update } from 'azle';
+
+type CertifiedValue = {
+    value: number;
```
Comment (lastmjs) [inline tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/src/index.ts:9]: TypeScript type for value and IDL do not match

File: tests/property/ic_api/certified_data/src/index.ts [hunk 37]
```diff
@@ -10,6 +10,16 @@ import {
     update
 } from 'azle';
 
+type CertifiedData = {
+    data: Uint8Array;
+    certificate: [Uint8Array] | [];
+};
+
+const CertifiedData = IDL.Record({
```
Comment (lastmjs) [inline tests/property/ic_api/certified_data/src/index.ts:13]: Please put IDL object definition above the typescript type

File: tests/property/ic_api/certified_data/jest.config.js [hunk 38]
```diff
@@ -0,0 +1,12 @@
+/** @type {import('ts-jest').JestConfigWithTsJest} */
+module.exports = {
+    preset: 'ts-jest',
+    testEnvironment: 'node',
+    bail: true,
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply tests/property/ic_api/certified_data/jest.config.js:5]: Were these removed?

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: tests/property/ic_api/certified_data/jest.config.js (hunk 38 above)
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/jest.config.js]: yes

File: .eslintrc.cjs
Comment (bdemann) [reply .eslintrc.cjs:1]: done I think

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/src/index.ts (hunk 35 above)
Comment (bdemann) [reply tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/src/index.ts]: done

File: tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/src/index.ts (hunk 36 above)
Comment (bdemann) [reply tests/end_to_end/candid_rpc/class_syntax/motoko_examples/cert-var/src/index.ts:9]: done

File: tests/property/ic_api/certified_data/src/index.ts (hunk 37 above)
Comment (bdemann) [reply tests/property/ic_api/certified_data/src/index.ts:13]: done

**Review by lastmjs — APPROVED** [review]

**Pull request merged**